luaui/Widgets/cmd_quick_build_extractor.lua

```diff
@@ -0,0 +1,263 @@
+function widget:GetInfo()
+	return {
+		name = "Quick Build (mex/geo)",
+		desc = "Adds ability to quickly place or upgrade mex/geothermal by right clicking.",
+		author = "Hobo Joe, based on work by Google Frog, NTG, Chojin, Doo, Floris, and Tarte",
+		version = "1.0",
+		date = "Jan 2024",
+		license = "GNU GPL, v2 or later",
+		handler = true,
+		layer = 1000,
+		enabled = true
+	}
+end
+
+local CMD_MOVE = CMD.MOVE
+local CMD_GUARD = CMD.GUARD
+local CMD_RECLAIM = CMD.RECLAIM
+
+local spGetActiveCommand = Spring.GetActiveCommand
+local spGetUnitDefID = Spring.GetUnitDefID
+local spGetUnitPosition = Spring.GetUnitPosition
+
+local geoPlacementRadius = 5000	-- (not actual ingame distance)
+local mexPlacementRadius = 2000	-- (not actual ingame distance)
+
+local mexConstructors
+local geoConstructors
+local mexBuildings
+local geoBuildings
+
+local bestGeo
+local bestMex
+
+local unitShape = false
+local activeShape
+
+local selectedSpot
+local selectedPos
+local selectedMex
+local selectedGeo
+
+local metalMap = false
+local buildCmd = {}
+
+local updateTime = 0
+
+local isCloakableBuilder = {}
+for unitDefID, unitDef in pairs(UnitDefs) do
+	if unitDef.buildOptions[1] and unitDef.canCloak then
+		isCloakableBuilder[unitDefID] = true
+	end
+end
+local function unitIsCloaked(uDefId)
+	return isCloakableBuilder[Spring.GetUnitDefID(uDefId)] and select(5,Spring.GetUnitStates(uDefId,false,true))
+end
+
+
+function widget:Initialize()
+	if not WG.DrawUnitShapeGL4 then
+		widget:Shutdown()
+	end
+
+	mexConstructors = WG["resource_spot_builder"].GetMexConstructors()
+	geoConstructors = WG["resource_spot_builder"].GetGeoConstructors()
+
+	mexBuildings = WG["resource_spot_builder"].GetMexBuildings()
+	geoBuildings = WG["resource_spot_builder"].GetGeoBuildings()
+
+	local metalSpots = WG["resource_spot_finder"].metalSpotsList
+	if not metalSpots or (#metalSpots > 0 and #metalSpots <= 2) then
+		metalMap = true
+	end
+end
+
+
+local function clearGhostBuild()
+	if WG.DrawUnitShapeGL4 and activeShape then
+		WG.StopDrawUnitShapeGL4(activeShape)
+		activeShape = nil
+	end
+	selectedMex = nil
+	selectedGeo = nil
+	selectedSpot = nil
+	selectedPos = nil
+	buildCmd = {}
+end
+
+
+local selectedUnits = Spring.GetSelectedUnits()
+function widget:SelectionChanged(sel)
+	selectedUnits = sel
+	bestMex = WG['resource_spot_builder'].GetBestExtractorFromBuilders(selectedUnits, mexConstructors, mexBuildings)
+	bestGeo = WG['resource_spot_builder'].GetBestExtractorFromBuilders(selectedUnits, geoConstructors, geoBuildings)
+end
+
+
+function widget:Update(dt)
+	if(selectedSpot) then
+		-- we want to do this every frame to avoid cursor flicker. Rest of work can be on a timer
```

Comment:
doesn't that mean some other widget is continuously reverting the cursor?

Reply:
About 30 of them are. It's very annoying and probably worth a refactor of its own, but its outside the scope of this work.